### PR TITLE
feat(depinject): support for core

### PR DESCRIPTION
**Pull request by crodriguezvega**
## Description

Adding support for `ibc` package.

ref: #3560

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [x] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -20,6 +20,7 @@ COPY Makefile .
 
 COPY go.mod .
 COPY go.sum .
+COPY ./api /go/api
```
Comment (crodriguezvega) [inline Dockerfile:23]: I had to do this to solve [this issue](https://github.com/cosmos/ibc-go/actions/runs/8256316313/job/22584782319) in the docker build.

File: proto/ibc/core/module/v1/module.proto [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import "cosmos/app/v1alpha1/module.proto";
+
+// Module is the config object of the core module.
+message Module {
+  option (cosmos.app.v1alpha1.module) = {
+    go_import: "github.com/cosmos/ibc-go/modules/core"
+  };
+
+  // authority defines the custom module authority. If not set, defaults to the governance module.
+  string authority = 1;
```
Comment (damiannolan) [inline proto/ibc/core/module/v1/module.proto:14]: any idea why this is on the proto definition? and where it gets set?

File: modules/core/depinject.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+var _ depinject.OnePerModuleType = AppModule{}
+
+// IsOnePerModuleType implements the depinject.OnePerModuleType interface.
+func (AppModule) IsOnePerModuleType() {}
+
+func init() {
+	appmodule.Register(
+		&modulev1.Module{},
+		appmodule.Provide(ProvideModule),
+	)
```
Comment (damiannolan) [inline modules/core/depinject.go:28]: I'm thinking that we may need something like `InvokeAddRoutes` here which will register `[]porttypes.IBCModule` on the router..

Something similar to x/gov maybe: https://github.com/cosmos/cosmos-sdk/blob/main/x/gov/depinject.go#L100

Hopefully we can use this as some kind of example. Happy to work on this once the initial PRs are merged.

File: proto/ibc/core/module/v1/module.proto (hunk 2 above)
Comment (damiannolan) [reply proto/ibc/core/module/v1/module.proto:14]: In `appmodule.Register` we pass an empty `&modulev1.Module{}` - same for sdk modules maintained over there afaik

**Review by damiannolan — APPROVED** [review]
Nice work! Thanks for keeping these PRs clean.

**Comment by crodriguezvega** [comment]
I have downgraded the `api` module to use go 1.21 to avoid all kind of build issues. We can upgrade the whole repo to 1.22 when we tackle #5776.

**Comment by crodriguezvega** [comment]
> I have downgraded the `api` module to use go 1.21 to avoid all kind of build issues. We can upgrade the whole repo to 1.22 when we tackle #5776.

Nah, it doesn't work.

**Comment by crodriguezvega** [comment]
> nice @crodriguezvega! Forgot ics23's `api/`!

hehe, nice, it's cool that it builds. 😅

**Comment by crodriguezvega** [comment]
@DimitrisJim If the CI is green and you approve this one, I can rebase the other PRs and hopefully we can get them merged quickly.

File: api/ibc/applications/fee/v1/query_grpc.pb.go
[the remark this answers is not in the document]
Comment (crodriguezvega) [reply api/ibc/applications/fee/v1/query_grpc.pb.go:1]: hmm, strange. I don't remember these being generated before... Let me check.

[the remark this answers is not in the document]
Comment (crodriguezvega) [reply api/ibc/applications/fee/v1/query_grpc.pb.go:1]: Indeed, I re-run `make proto-all` and the disappeared.

**Pull request merged**